types/propEq.d.ts

```diff
@@ -1,6 +1,32 @@
-export function propEq<T>(val: T): {
-  <K extends PropertyKey>(name: K): (obj: Record<K, T>) => boolean;
-  <K extends PropertyKey>(name: K, obj: Record<K, T>): boolean;
+import { Placeholder } from 'ramda';
+import { WidenLiterals } from '../util/tools';
+
+export function propEq(__: Placeholder): never;
+export function propEq<V>(val: V): {
+  <K extends PropertyKey>(name: K): K extends number
+    ? (array: Array<WidenLiterals<V>>) => boolean
+    : (obj: Record<K, WidenLiterals<V>>) => boolean;
+  <const U>(__: Placeholder, obj: U): (name: keyof U) => boolean;
+  <K extends keyof U, const U>(name: K, obj: U): boolean;
 };
-export function propEq<T, K extends PropertyKey>(val: T, name: K): (obj: Record<K, T>) => boolean;
-export function propEq<K extends keyof U, U>(val: U[K], name: K, obj: U): boolean;
+export function propEq<K extends PropertyKey>(__: Placeholder, name: K): K extends number ? {
```

Comment:
Instead of the ternary, we do 2 overloads
```typescript
export function propEq<K extends number>(__: Placeholder, name: K): {
  // ...
}
export function propEq<K extends Exclude<PropertyKey, number>>(__: Placeholder, name: K): {
  // ...
}
```